plat/common/lcpu.c

```diff
@@ -48,3 +165,424 @@ void ukplat_lcpu_halt_to(__nsec until)
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for wakeup IRQ %d\n",
+			   LCPU_ARCH_IPI_WAKEUP_IRQ);
+		return rc;
+	}
+
+	return 0;
+}
+
+void __weak __noreturn lcpu_entry_default(struct lcpu *this_lcpu)
+{
+	int rc;
+
+	/* Finish initialization. As there is nothing to return to, we
+	 * just enter halted state if an error occurs.
+	 */
+	rc = lcpu_init(this_lcpu);
+	if (unlikely(rc))
+		lcpu_halt(this_lcpu, rc);
+
+	/* We are coming from BUSY0 state and want to transition to IDLE
+	 * state. However, there can already be functions queued so we have
+	 * to use a decrement here.
+	 */
+	ukarch_dec(&this_lcpu->state);
+
+	/* Enable IRQs. If there are functions queued we will immediately
+	 * jump to the IRQ handler.
+	 */
+	ukplat_lcpu_enable_irq();
+	while (1) {
+		/* Besides interrupts in general, the halt can be
+		 * interrupted by calls to ukplat_lcpu_run().
+		 */
+		halt();
+	}
+}
+
+int ukplat_lcpu_start(const __lcpuidx lcpuidx[], unsigned int *num, void *sp[],
+		      const ukplat_lcpu_entry_t entry[], unsigned long flags)
+{
+	__lcpuid this_cpu_id = ukplat_lcpu_id();
+	struct lcpu *lcpu;
+	unsigned int i, n, argi = 0;
+	const int new = LCPU_STATE_INIT;
+	int old;
+	int rc = 0, rc2 = 0;
+
+	UK_ASSERT(((lcpuidx) && (num)) || ((!lcpuidx) && (!num)));
+	UK_ASSERT(sp);
+
+	lcpu_lcpuidx_list_foreach(lcpuidx, num, n, i, lcpu) {
+		if (lcpu->id == this_cpu_id) {
+			/* If the caller did not supply an index array, we
+			 * assume that we do not have parameters and a stack
+			 * for the executing CPU. Otherwise, i.e., if the
+			 * caller explicitly put the executing CPU in, we still
+			 * ignore it but need to skip the parameters.
+			 */
+			if (lcpuidx)
+				argi++;
+
+			continue;
+		}
+
+retry:
+		old = ukarch_load_n(&lcpu->state);
+
+		/* We ignore CPUs that are already started */
+		if (unlikely(old != LCPU_STATE_OFFLINE)) {
+			uk_pr_warn("Failed to start CPU 0x%lx: not offline\n",
+				   lcpu->id);
+
+			/* Skip CPU and its arguments*/
+			argi++;
+			continue;
+		}
+
+		/* Try to acquire the CPU for initialization. If another thread
+		 * was faster, we will return to the state comparison and
+		 * report that the CPU is not offline.
+		 */
+		if (ukarch_compare_exchange_sync((int *)&lcpu->state, old,
+						 new) != new)
+			goto retry;
+
+		UK_ASSERT(lcpu->state == LCPU_STATE_INIT);
+
+		/* Setup startup arguments.
+		 * Since we are ignoring the executing CPU, we must keep a
+		 * separate counter to index the arguments.
+		 */
+		lcpu->s_args.entry = (entry && entry[argi]) ?
+			entry[argi] : lcpu_entry_default;
+		lcpu->s_args.stackp = sp[argi];
+
+		/* Ensure that the startup arguments have been written back
+		 * before issuing the startup call
+		 */
+		wmb();
+
+		rc = lcpu_arch_start(lcpu, flags);
+		if (unlikely(rc)) {
+			lcpu->state = LCPU_STATE_HALTED;
+			lcpu->error_code = rc;
+
+			/* There is a more serious problem. Stop here. The
+			 * caller can skip the CPU and retry if the likes to
```

Comment:
Something's wrong with that sentence...

Reply:
👍🏻

Reply:
I guess he wanted to say "if he likes to", but I would replace it with "if it wants to"

Reply:
Right, the sentence got messed up. Actually, at the moment you cannot even retry because the CPU is put into `HALTED` state. As we discussed, we have to think more about how to handle errors at this level. Perhaps have a reset function or so. For now, we at least inform about the error.